File: node/import-matrix.py
#! /usr/bin/env python3

import asyncio
import csv
import gzip
import logging
from dataclasses import dataclass
from os import path, walk, environ
import re
from shutil import copyfile
from sys import argv
from typing import List, Dict, Any
import time

from motor import motor_asyncio

client = motor_asyncio.AsyncIOMotorClient(environ.get("MONGO_CONNECTION_STRING"))
db = client[environ.get("MONGO_DB")]
features_collection = db.features

logger = logging.getLogger(__name__)


def configure_logger(debug = False):
    logger.setLevel(logging.DEBUG if debug else logging.INFO)
    ch = logging.StreamHandler()
    formatter = logging.Formatter('%(asctime)s - %(message)s')
    ch.setFormatter(formatter)
    logger.addHandler(ch)


@dataclass
class FileSet:
    matrix_files: bool = False
    tree: bool = False
    labels: bool = False

    def all_found(self):
        return self.matrix_files and self.tree and self.labels

    def __str__(self):
        if self.all_found():
            pass
        else:
            missing = ''
            if not self.matrix_files:
                missing += " \nMatrix files (features.tsv or genes.tsv, barcodes.tsv, matrix.mtx, all in one directory or subdirectory). Do they need to be extracted?"
            if not self.tree:
                missing += " \ncluster_tree.json"
            if not self.labels:
                missing += "\nlabels.csv"

            return f"The import script failed to find the following files in the data directory: {missing}"


async def insert_many(docs: List[Dict[Any, Any]]):
    await features_collection.insert_many(docs)


async def bootstrap():
    await features_collection.drop()


start = time.perf_counter()


ex = re.compile(r"^(features\.tsv|genes\.tsv|barcodes\.tsv|matrix\.mtx)(\.gz)?$")
async def contains_matrix_files(files: List[str]):
    """ returns true if matrix files exist in file list """
    return len(set([f for f in files if ex.match(f)])) == 3

async def check_for_files(root_dir: str):
    """ this will run contains_files and die once everything is found """
    results = FileSet()
    for root, dirs, files in walk(root_dir):
        if await contains_matrix_files(files):
            results.matrix_files = True

        if "cluster_tree.json" in files:
            results.tree = True

        if "labels.csv" in files:
            results.labels = True

        if results.all_found():
            break

    if results.all_found():
        return True
    else:
        raise FileNotFoundError(results)

async def open_maybe_compressed(pathname: str):
    if pathname.endswith('.gz'):
        return gzip.open(pathname,  mode="rt")
    else:
        return open(pathname,  mode="rt")


async def parse_matrices(root_dir: str):
    for root, dirs, files in walk(root_dir):
        if await contains_matrix_files(files):
            matrix_file = next(f for f in files if re.match(r"^matrix.mtx",f))
            with await open_maybe_compressed(path.join(root, matrix_file)) as f:
                logger.info(f"loading {root}")
                feature_file = next(f for f in files if re.match(r"^(features|genes).tsv",f))
                features_path = path.join(root, feature_file)
                feature_ids = {
                    i
                    + 1: {
                        "feature": row[0],
                        "feature_type": row[1]
                        if len(row) > 1 and row[1] != row[0]
                        else "Gene Expression",
                    }
                    for i, row in enumerate(
                        csv.reader(await open_maybe_compressed(features_path), delimiter="\t")
                    )
                }
                barcodes_file = feature_file = next(f for f in files if re.match(r"^barcodes.tsv",f))
                barcodes_path = path.join(root, barcodes_file)
                barcodes = {
                    i + 1: row[0]
                    for i, row in enumerate(
                        csv.reader(await open_maybe_compressed(barcodes_path), delimiter="\n")
                    )
                }
                i = 0
                j = 1
                chunk = []
                #skip headers
                [f.readline() for i in range(2)]
                for line in f:
                    feature_idx, barcode_idx, value = line.split(" ")
                    chunk.append(
                        {
                            "feature": feature_ids[int(feature_idx)]["feature"],
                            "feature_type": feature_ids[int(feature_idx)]["feature_type"],
                            "id": barcodes[int(barcode_idx)],
                            "value": float(value),
                        }
                    )
                    i+=1
                    if i == 100000:  # 100k
                        logger.debug(f"inserting chunk {j}...")
                        await insert_many(chunk)
                        logger.debug(f"chunk {j} inserted.")
                        i = 0
                        j += 1
                        chunk = []
                #insert remaining rows
                await insert_many(chunk)

        if "cluster_tree.json" in files:
            copyfile(
                path.join(root, "cluster_tree.json"),
                "/usr/app/static/files/cluster_tree.json",
            )

        if "labels.csv" in files:
            copyfile(path.join(root, "labels.csv"), "/usr/app/static/files/labels.csv")

async def insert_features(debug: bool):
    configure_logger(debug=debug)
    logger.debug("debugging enabled!")
    await bootstrap()
    logger.info("checking that required files are in the data directory...")
    await check_for_files("/usr/data")
    logger.info("importing files into mongo...")
    await parse_matrices("/usr/data")
    logger.info("building index...")
    await features_collection.create_index("feature")
    logger.info(f"finished import in {time.perf_counter() - start} seconds")


if __name__ == "__main__":

    debug = True if (len(argv) > 1 and argv[1] == "--debug") else False

    asyncio.run(insert_features(debug))

File: node/scripts/importMatrix.ts
import { opendir } from 'fs/promises';
import { createGunzip } from 'zlib';
import { existsSync, createReadStream } from 'fs';
import path from 'path';

import split2 from 'split2';
import yargs from 'yargs';
import { Pool } from 'pg';
import { from as copyFrom } from 'pg-copy-streams';

interface Feature {
    feature: string;
    feature_type: string;
}

const argv = yargs(process.argv.slice(2))
    .option('debug', {
        describe: 'Provide verbose output',
    })
    .help().argv as Record<string, string>;

const DEBUG = !!argv.debug;

const CHUNK_SIZE = 100000;

const debug = (msg: string) => {
    if (argv.debug) {
        console.debug(`[${new Date().toLocaleTimeString()}] ${msg}`);
    }
};

const getPool = async (remainingTries = 5) => {
    let client;
    let pool;
    try {
        pool = new Pool();
        client = await pool.connect();
    } catch (e) {
        console.error(e);
        if (remainingTries) {
            await new Promise(resolve =>
                setTimeout(() => resolve(getPool(remainingTries - 1)), 1000)
            );
        } else {
            throw 'Could not connect to database!';
        }
    } finally {
        client?.release();
    }
    return pool as Pool;
};

const verifyMatrixFilesPresent = async (rootDir: string) => {
    const missingFiles = {
        matrixFiles: 'matrix files not found',
    } as Partial<Record<string, string>>;

    for await (const filelist of walkTree(rootDir)) {
        if (containsMatrixFiles(filelist)) {
            delete missingFiles.matrixFiles;
        }
    }

    if (Object.values(missingFiles).length) {
        throw Object.values(missingFiles).join('\n');
    }
};

const truncateDatabase = async (pool: Pool) => {
    const client = await pool.connect();
    await client.query('TRUNCATE features;');
    await client.query('DROP INDEX IF EXISTS feature_idx;');
    client.release();
};

const createIndex = async (pool: Pool) => {
    const client = await pool.connect();
    await client.query(
        'CREATE INDEX IF NOT EXISTS feature_idx ON features (feature);'
    );
    client.release();
};

const insertMany = async (chunk: string, pool: Pool) => {
    const query = 'COPY features (id, feature, feature_type, value) FROM STDIN';

    const client = await pool.connect();

    const stream = client.query(copyFrom(query));

    stream.write(Buffer.from(chunk), err => {
        if (err) {
            console.error(err);
        }
        stream.end();
    });

    return new Promise((resolve, reject) => {
        stream.on('error', e => {
            client.release();
            reject(e);
        });
        stream.on('finish', e => {
            client.release();
            resolve(true);
            stream.destroy();
        });
    });
};

/**
 *
 * @param filepath Path to features file
 * @returns A dictionary of features keyed by index for fast lookup
 */
const getFeaturesDict = async (filepath: string) => {
    const ret = {} as Record<string, Feature>;
    let i = 1;
    await parseFile(filepath, async row => {
        const vals = row.split('\t');
        ret[i] = {
            feature: vals[0],
            feature_type:
                vals.length > 1 && vals[1] != vals[0]
                    ? vals[1]
                    : 'Gene Expression',
        };
        i++;
    });

    return ret;
};

/**
 *
 * @param filepath Path to barcode file
 * @returns A dictionary of barcodes keyed by index for fast lookup
 */
const getBarcodesDict = async (filepath: string) => {
    const ret = {} as Record<string, string>;
    let i = 1;
    await parseFile(filepath, async row => {
        ret[i] = row.split('\t')[0];
        i++;
    });

    return ret;
};

/**
 *
 * @param filepath Location of file to open
 * @param cb callback to fire on 'data' event
 * @returns Promise<boolean>
 */
const parseFile = async (
    filepath: string,
    cb: (line: string) => Promise<void>
): Promise<boolean> =>
    new Promise(async (resolve, reject) => {
        const rl = openMaybeCompressed(filepath);

        rl.on('data', async line => {
            rl.pause();
            await cb(line);
            rl.resume();
        })
            .on('finish', () => {
                resolve(true);
            })
            .on('error', e => {
                reject(e);
            });
    });

/**
 * Open a read stream and pipe to split2 in order to read line-by-line
 *
 * @param filepath Path of the file to open
 * @returns Transform
 */
const openMaybeCompressed = (filepath: string) => {
    return filepath.endsWith('gz')
        ? createReadStream(filepath).pipe(createGunzip()).pipe(split2())
        : createReadStream(filepath).pipe(split2());
};

/**
 * Import the matrix by reading row by row and attaching values from barcodes and features files
 *
 * @param barcodes Dictionary of barcodes by index
 * @param features Dictionary of features by index
 * @param matrixFilepath Path to matrix
 * @param lineCount Optional line count for debugging
 * @returns
 */
const insertMatrix = async (
    pool: Pool,
    barcodes: Record<string, string>,
    features: Record<string, Feature>,
    matrixFilepath: string,
    lineCount?: number
) => {
    let i = 0;
    let j = 0;
    let chunk = '';
    await parseFile(matrixFilepath, async line => {
        //skip headers
        if (i < 3 && j === 0) {
            i++;
            return;
        }
        const [featureIdx, barcodeIdx, value] = line.split(' ');

        chunk += `${barcodes[barcodeIdx]}\t${features[featureIdx].feature}\t${features[featureIdx].feature_type}\t${value}\n`;

        i++;
        if (i === CHUNK_SIZE) {
            if (lineCount) {
                debug(
                    `Inserting chunk ${j} (${(
                        (((j + 1) * CHUNK_SIZE) / lineCount) *
                        100
                    ).toFixed(2)}%)`
                );
            }

            await insertMany(chunk, pool);
            debug(`Chunk ${j} inserted.`);
            i = 0;
            j++;
            chunk = '';
        }
    });
    await insertMany(chunk, pool);

    return true;
};

const countLines = async (filepath: string): Promise<number> => {
    let count = 0;
    await parseFile(filepath, async () => {
        count++;
    });

    return count;
};

/* Walk the tree, yielding full contents of each directory */
async function* walkTree(root: string): AsyncGenerator<string[], any, any> {
    const files = [] as string[];
    const dirs = [] as string[];
    for await (const d of await opendir(root)) {
        const entry = path.join(root, d.name);
        if (d.isDirectory()) {
            dirs.push(entry);
        } else if (d.isFile()) {
            files.push(entry);
        }
    }

    yield files;
    if (dirs.length) {
        for (const dir of dirs) {
            yield* walkTree(dir);
        }
    }
}

const mtxRegex = /(features\.tsv|genes\.tsv|barcodes\.tsv|matrix\.mtx)(\.gz)?$/;

const containsMatrixFiles = (files: string[]) =>
    files.filter(f => mtxRegex.test(f)).length === 3;

const importData = async (rootDir: string, pool: Pool) => {
    if (!existsSync(rootDir) || !path.isAbsolute(rootDir)) {
        throw `${rootDir} does not exist! Is the path absolute?`;
    }

    for await (const filelist of walkTree(rootDir)) {
        if (containsMatrixFiles(filelist)) {
            const matrixFile = filelist.find(f =>
                f.match(/matrix\.mtx(\.gz)?$/)
            )!;
            let lineCount;
            if (DEBUG) {
                lineCount = await countLines(matrixFile);
                debug(`Inserting ${lineCount} rows from ${matrixFile}`);
            }

            const featureFile = filelist.find(f =>
                /(features|genes)\.tsv(\.gz)?$/.test(f)
            )!;

            const features = await getFeaturesDict(featureFile);

            const barcodesFile = filelist.find(f =>
                /barcodes.tsv(\.gz)?$/.test(f)
            )!;

            const barcodes = await getBarcodesDict(barcodesFile);

            await insertMatrix(pool, barcodes, features, matrixFile, lineCount);
        }
    }
};

/**
 * Import files into postgres.
 *  Note that the files are expected to be located in the `/usr/data` directory.
 *  As this script is expected to be executed as part of a `docker run` command, the user should
 *  bind-mount the needed files into the /usr/data directory.
 */
const run = async () => {
    const dataDir = '/usr/data';
    const pool = await getPool();
    console.log('Data import started');
    debug('verifying data');
    await verifyMatrixFilesPresent(dataDir);
    debug('truncating database');
    await truncateDatabase(pool);
    debug('uploading files');
    const start = new Date().getTime();
    await importData(dataDir, pool);
    debug('creating index');
    await createIndex(pool);
    const end = new Date().getTime();
    console.log(`finished in ${(end - start) / 1000} seconds`);
};

run();
